### PR TITLE
Simple CMake setup for ONNX hosting

### PR DESCRIPTION
**Pull request by tmccrmck**
- Adds to build.sh and build.py
- Simple CMake integration to link with inference_session

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -65,6 +65,7 @@ option(onnxruntime_ENABLE_MICROSOFT_INTERNAL "Use this option to enable/disable
 option(onnxruntime_USE_NUPHAR "Build with Nupha" OFF)
 option(onnxruntime_USE_BRAINSLICE "Build with BrainSlice" OFF)
 option(onnxruntime_USE_TRT "Build with TensorRT support" OFF)
+option(onnxruntime_BUILD_HOSTING "Build ONNX hosting service" ON)
```
Comment (NonStatic2014) [inline cmake/CMakeLists.txt]: Better leave a TODO here make sure before PR to upstream master, we make it default `OFF`. Also, we are not a service, just an application. We could just call it "ONNX hosting".

File: cmake/onnxruntime_hosting.cmake [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+set(CMAKE_CXX_STANDARD 14)
+
+include_directories(${PROJECT_NAME}" PRIVATE ${ONNXRUNTIME_ROOT}/hosting/include")
+
+file(GLOB_RECURSE onnxruntime_hosting_srcs
+    "${ONNXRUNTIME_ROOT}/hosting/*.h"
+    "${ONNXRUNTIME_ROOT}/hosting/*.cc"
+)
+
+source_group(TREE ${REPO_ROOT} FILES ${onnxruntime_hosting_srcs})
```
Comment (NonStatic2014) [inline cmake/onnxruntime_hosting.cmake:17]: Is this for IDE only?

File: onnxruntime/hosting/main.cc [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#include "core/session/inference_session.h"
```
Comment (NonStatic2014) [inline onnxruntime/hosting/main.cc:4]: Missing "copyright" header. For all new files.

File: tools/ci_build/build.py [hunk 4]
```diff
@@ -87,6 +87,9 @@ def parse_arguments():
     # Build a shared lib
     parser.add_argument("--build_shared_lib", action='store_true', help="Build a shared library for the ONNXRuntime.")
 
+    # Build a shared lib
```
Comment (NonStatic2014) [inline tools/ci_build/build.py]: Build ONNX hosting

File: tools/ci_build/build.py [hunk 5]
```diff
@@ -87,6 +87,9 @@ def parse_arguments():
...
+    parser.add_argument("--build_hosting", action='store_true', help="Build HTTP hosting binary for the ONNXRuntime.")
```
Comment (NonStatic2014) [inline tools/ci_build/build.py]: Don't have to be HTTP only. :-) you could just call it hosting application.

File: tools/ci_build/build.py [hunk 6]
```diff
@@ -328,6 +332,8 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
 
     cmake_args += ["-D{}".format(define) for define in cmake_extra_defines]
 
+    cmake_args += ["-DCMAKE_TOOLCHAIN_FILE=/home/trmccorm/vcpkg/scripts/buildsystems/vcpkg.cmake"]
```
Comment (NonStatic2014) [inline tools/ci_build/build.py]: Better make it a relative path to vcpkg in the`hosting` folder.

File: cmake/onnxruntime_hosting.cmake (hunk 2 above)
Comment (tmccrmck) [reply cmake/onnxruntime_hosting.cmake:17]: For VS: https://cmake.org/cmake/help/v3.12/command/source_group.html?highlight=s

File: onnxruntime/hosting/main.cc (hunk 3 above)
Comment (tmccrmck) [reply onnxruntime/hosting/main.cc:4]: Done.

File: tools/ci_build/build.py [hunk 7]
```diff
@@ -199,6 +202,15 @@ def install_python_deps():
     dep_packages = ['setuptools', 'wheel', 'numpy']
     run_subprocess([sys.executable, '-m', 'pip', 'install', '--trusted-host', 'files.pythonhosted.org'] + dep_packages)
 
+def install_hosting_deps():
```
Comment (NonStatic2014) [inline tools/ci_build/build.py]: This function should accept something like "`source_dir`" as parameter and use it as the first parameter in `os.path.join()`. Check other functions in this file as example.

File: tools/ci_build/build.py [hunk 8]
```diff
@@ -328,6 +341,9 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
 
     cmake_args += ["-D{}".format(define) for define in cmake_extra_defines]
 
+    toolchain_path = os.path.join('cmake', 'external', 'vcpkg', 'scripts', 'buildsystems', 'vcpkg.cmake')
```
Comment (NonStatic2014) [inline tools/ci_build/build.py]: Better replace `'cmake'` with `'cmake_dir'`

File: tools/ci_build/build.py [hunk 9]
```diff
@@ -328,6 +341,9 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
...
+    cmake_args += ["-DCMAKE_TOOLCHAIN_FILE=" + toolchain_path]
```
Comment (NonStatic2014) [inline tools/ci_build/build.py]: It is OK for now. But if any CMakeLists.txt using `find_package()` and vcpkg happens to have that package, then the build end up using the vcpkg package other than the package in the repo. You could set the `CMAKE_TOOLCHAIN_FILE` in the hosting.cmake file with an environment variable. In this file you just set the environment variable.

File: tools/ci_build/build.py (hunk 7 above)
Comment (tmccrmck) [reply tools/ci_build/build.py]: Done.

File: tools/ci_build/build.py (hunk 9 above)
Comment (tmccrmck) [reply tools/ci_build/build.py]: Done

File: tools/ci_build/build.py (hunk 8 above)
Comment (tmccrmck) [reply tools/ci_build/build.py]: Done

File: cmake/onnxruntime_hosting.cmake [hunk 10]
```diff
@@ -0,0 +1,40 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+if(NOT DEFINED CMAKE_TOOLCHAIN_FILE)
+    set(CMAKE_TOOLCHAIN_FILE "${ONNXRUNTIME_ROOT}/cmake/external/vcpkg/scripts/buildsystems/vcpkg.cmake" CACHE STRING "")
+endif()
+
+project(onnxruntime_hosting)
+set(CMAKE_CXX_STANDARD 14)
```
Comment (pranavsharma) [inline cmake/onnxruntime_hosting.cmake]: this gets set at the start of CMakeLists.txt, so it might be redundant.

**Review by pranavsharma — APPROVED** [review]

File: cmake/onnxruntime_hosting.cmake (hunk 10 above)
Comment (tmccrmck) [reply cmake/onnxruntime_hosting.cmake]: Fixed.

**Pull request merged**